### PR TITLE
Chapter2/Homework/2

### PR DESCRIPTION
**Pull request by matthew-hole**
First commit. 

PR contains the solution to meet the listed requirements (Unit Tests), but does not include the 'Extra Challenge: Equipment' functionality which will be implemented at a later date.

Only the ChestPiece class was refactored into an 'Item' child class to demonstrate the functionality. A test was performed in the 'Program.cs' file where the functionality is tested.

In the next commit, the Equipment Functionality will be implemented alongside a 'Player/Shop Simulator' which will test all core functionality with edge cases and robustness testing.

### CODE REVIEW

File: Src/BootCamp.Chapter/Inventory.cs [hunk 1]
```diff
@@ -1,26 +1,62 @@
...
 {
     public class Inventory
     {
-        private Item[] _items;
-        public Item[] GetItems()
+        protected Item[] _items;
+
+        // displays all items in inventory
+        // renamed from 'GetItems' to show the clear different from the 'GetItems' method below
+        public void ShowAllItems()
```
Comment (Almantask) [inline Src/BootCamp.Chapter/Inventory.cs:13]: It doesn't show all items, it displays them; try to be more conventional.

File: Src/BootCamp.Chapter/Inventory.cs [hunk 2]
```diff
@@ -1,26 +1,62 @@
...
         {
-            return new Item[0];
+            if (!Array.TrueForAll(_items, IsNull => IsNull == null))
```
Comment (Almantask) [inline Src/BootCamp.Chapter/Inventory.cs:15]: What if array is empty?

Comment (Almantask) [reply Src/BootCamp.Chapter/Inventory.cs:15]: Try to fail early and avoid negations.
If there are no elements- end the function.
if(Array.TrueForAll(_items, IsNull => IsNull == null))
{
  Console.WriteLine("The inventory is currently empty.");
  return;
}
// no need else. Reduced nesting.

File: Src/BootCamp.Chapter/Inventory.cs [hunk 3]
```diff
@@ -1,26 +1,62 @@
...
+            {
+                Console.WriteLine("Items in inventory:\n");
+                foreach (var item in _items)
+                {
+                    if (item != null)
```
Comment (Almantask) [inline Src/BootCamp.Chapter/Inventory.cs:20]: Invert the if.
```
if(item == null) continue;
Console.WriteLine(item.GetName())
```

File: Src/BootCamp.Chapter/Inventory.cs [hunk 4]
```diff
@@ -1,26 +1,62 @@
...
+            {
+                Console.WriteLine("The inventory is currently empty.");
+            }
         }
 
-        public Inventory()
+        public Item GetItem(string name)
         {
-            _items = new Item[0];
+            return _items[Array.FindIndex(_items, element => element.GetName().Equals(name))];
```
Comment (Almantask) [inline Src/BootCamp.Chapter/Inventory.cs:38]: It does work, but in general you should treat C# as raw, without all the neat language features.

File: Src/BootCamp.Chapter/Player.cs [hunk 5]
```diff
@@ -28,34 +34,55 @@ public class Player
         /// <summary>
         /// Player items. There can be multiple of items with same name.
         /// </summary>
-        private Inventory _inventory;
+        private PlayerInventory _inventory;
```
Comment (Almantask) [inline Src/BootCamp.Chapter/Player.cs:37]: Why have a distinction between just inventory? Why is player inventory is any special? Keep it simple, don't be redundant.

File: Src/BootCamp.Chapter/Player.cs [hunk 6]
```diff
@@ -28,34 +34,55 @@ public class Player
...
+          
         }
 
         /// <summary>
         /// Adds item to player's inventory
         /// </summary>
+        // The inventory method 'AddItem' will throw an expection if no room is available in the inventory for an another item
         public void AddItem(Item item)
         {
+            try
```
Comment (Almantask) [inline Src/BootCamp.Chapter/Player.cs:73]: It's not that excpetional that inventory is full. Instead, check the inventory size before adding more items to it. throwing an exception is costly and it should also be reserved for unexpected situations.

**Review by Almantask — APPROVED** [review]
Overall, looks good.
